extensions/cognigy/src/knowledge-connectors/chunker/simpleSplit.ts

```diff
@@ -0,0 +1,70 @@
+/* Node Modules */
+import { TokenTextSplitter, TokenTextSplitterParams, CharacterTextSplitter, CharacterTextSplitterParams } from "langchain/text_splitter";
+import * as modelToEncoding from '@dqbd/tiktoken/model_to_encoding.json';
+import { TiktokenEncoding, TiktokenModel } from "@dqbd/tiktoken";
+
+// Max chunk size limit
+const MAX_CHUNK_SIZE = 2000;
+
+/**
+ * This method resolves the proper encoding name for a given language model.
+ * It's a helper function to find the proper encoding name for a model without looking it up.
+ * The mapping comes from the "@dqbd/tiktoken" package, which is internally used by langchain
+ *
+ * @param modelName a tiktoken-compatible model name
+ * @returns a tiktoken-compatible encoding name
+ */
+const getEncodingNameForModelName = (modelName: TiktokenModel): TiktokenEncoding => {
+    const encodingName = modelToEncoding[modelName] as TiktokenEncoding;
+    if (!encodingName)
+        return null;
+    return encodingName;
+};
+
+interface ISimpleSplitOptions extends TokenTextSplitterParams {};
+const TokenSplitterDefaultOptions: Partial<ISimpleSplitOptions> = {
+    chunkSize: 512,
+    chunkOverlap: 128,
+    encodingName: getEncodingNameForModelName('gpt-3.5-turbo')
+};
+
+interface ITextSplitOptions extends CharacterTextSplitterParams {};
+const CharSplitDefaultOptions: Partial<ITextSplitOptions> = {
+    chunkSize: MAX_CHUNK_SIZE,
+    chunkOverlap: 0,
+};
+
+/**
+ * Splits a text into fixed-length chunks.
+ *
+ * First applies semantic splitting on Token level, then refines it using character level splitting
+ * in case the chunks are still greater than the MAX_CHUNK_SIZE.
+ *
+ * @param text a string that should be split into chunks
+ * @param options a set of options that should be passed to the TokenTextSplitter
+ * @returns
+ */
+export const simpleSplit = async (
+    text: string,
+    options: Partial<ISimpleSplitOptions> = {}
+): Promise<String[]> => {
+    const splitter = new TokenTextSplitter({
+        ...TokenSplitterDefaultOptions,
+        ...options
+    });
+    const chunks = await splitter.splitText(text);
+    const charSplitter = new CharacterTextSplitter({
+        ...CharSplitDefaultOptions
+    });
+
+    let refinedChunks: string[] = [];
+    for (const chunk of chunks) {
+        if (chunk.length > MAX_CHUNK_SIZE) {
+            const smallerChunks = await charSplitter.splitText(chunk);
+            refinedChunks = refinedChunks.concat(smallerChunks);
+        } else {
+            refinedChunks.push(chunk);
+        }
+    }
```

Comment:
This block handles making chunks smaller than `MAX_CHUNK_SIZE` by splitting the chunks again. Is it possible that chunks returned from `const chunks = await splitter.splitText(text);` have a bigger size than what is passed as `chunkSize`? If not, then it would be sufficient to just not allow chunkSize to be more than max size. Here it could be done via

```
    const splitter = new TokenTextSplitter({
        ...TokenSplitterDefaultOptions,
        ...options,
        chunkSize: options?.chunkSize ? Math.min(options.chunkSize, MAX_CHUNK_SIZE) : TokenSplitterDefaultOptions.chunkSize
    });
```

Then all of the refinedChunks block could be removed in favor of `return splitter.splitText(text);` (await is not needed)

Reply:
@falak-asad this comment is still open

Reply:
Got it, thank you for clarifying :+1: